File: PorizoAndroid/Sources/PorizoSkipSpike/PoemClaimLogic.swift
import Foundation

/// Pure poem-claim decision logic (U13). Reuses `ClaimLogic.State`; maps
/// poem-share status and extracts the verses to reveal. No UI, no network.
enum PoemClaimLogic {

    /// Map a poem-share response to a claim state (shares the track claim states).
    static func state(for info: PorizoPoemShareInfoResponse) -> ClaimLogic.State {
        switch info.status {
        case "unbound":
            return .claimable(needsPin: info.requiresPinForClaim ?? info.requiresPin ?? false)
        case "claimed":
            return .claimed
        default:
            return .unavailable
        }
    }

    /// The verses to display: full verses when present, else preview lines.
    static func verses(from info: PorizoPoemShareInfoResponse) -> [String] {
        verses(from: info.poem)
    }

    static func verses(from poem: PorizoPoemBody?) -> [String] {
        guard let poem else { return [] }
        if let verses = poem.verses, !verses.isEmpty { return verses }
        return poem.previewLines ?? []
    }
}

File: PorizoAndroid/Android/core/domain/src/main/kotlin/com/porizo/core/domain/claim/PoemClaimLogic.kt
package com.porizo.core.domain.claim

import com.porizo.core.model.PoemBody
import com.porizo.core.model.PoemShareInfo

object PoemClaimLogic {
    fun stateFor(info: PoemShareInfo): ClaimLogic.State =
        when (info.status) {
            "active", "unbound" -> ClaimLogic.State.Claimable(
                needsPin = info.requiresPinForClaim ?: info.requiresPin ?: false,
            )
            "claimed" -> ClaimLogic.State.Claimed
            else -> ClaimLogic.State.Unavailable
        }

    fun verses(info: PoemShareInfo): List<String> = verses(info.poem)

    fun verses(poem: PoemBody?): List<String> {
        if (poem == null) return emptyList()
        val verses = poem.verses
        if (!verses.isNullOrEmpty()) return verses
        return poem.previewLines.orEmpty()
    }
}
